### ISSUE TITLE
Redesign Consume with batches of records for reliable flow control

### ISSUE THREAD

**Issue by david-eckardt-sociomantic** [sociomantic-tsunami/dmqproto#3]
The current design of Consume throttling with suspend/resume control messages doesn't work well if the application has more swarm activity than just the Consumer request. This is a result from testing with a real consuming application and a reduced test case which per received message spawns a task that pushes the record to another DMQ node. Between sending the suspend message and receiving the ACK from the node the application receives umpteen thousand records. Yielding receive and send loops in the node and client or disabling socket output buffering doesn't help. Reducing the client socket input buffer size via `setsockopt(SOL_SOCKET, SO_RCVBUF)` does help but would slow down the whole socket connection.
Flow control for Consume needs to be done differently. We came up with the following idea:
- The node sends a batch of records with a certain maximum size, say, tens or thousands of kilobytes and waits for the client to request the next batch.
- The client Consume moves the received batch into a queue, then pops all records from it, calling the user notifier for each record. When the queue is empty -- the application has processed the batch -- then the client requests the next batch from the node.

The user notifying iteration over the received batch happens under the hood and is suspendable so that it stays compatible with existing throttling infrastructure. The advantage of this design is that the client never receives more records than it can process in time.
A disadvantage is that efficiency is sacrificed: Continuous feedback is involved, and there is a delay of at least one round-trip time between the batches. Overall this Consume design is not streaming any more, it is a sequence of "pop batch" requests instead. However, the efficiency of streaming Consume is that high that it overloads even simple test programs. With multiple nodes the throughput is even multiplied. When a batch from each of two nodes A and B are waiting to be processed by the application, the application first processes batch A, then the client requests the next batch from node A, but until that batch has arrived the application still needs to process batch B.
Another thing to watch out for is to avoid long delays in the node if records are sent at a low rate as it happens when a DMQ channel is empty and records are pushed to it at a low rate. In this situation aggregating a batch of record may take a long time so there should be a timeout after which the node sends an incomplete batch. For production one second should be enough, for testing much lower values may be required.

**Comment by gavin-norman-sociomantic** [comment]
>When a batch from each of two nodes A and B are waiting to be processed by the application, the application first processes batch A, then the client requests the next batch from node A, but until that batch has arrived the application still needs to process batch B.

Actually it'd probably be implemented like this:
```D
foreach ( record; this.received_batch )
{
  if ( user_suspended )
    waitForResume();

  notifyUser(record);
  periodicYield(10); // yield back to epoll after 10 records
}
```
So the records from each node's batch would be interleaved. (This is a really internal detail, though, so not that important at this stage.)

**Comment by gavin-norman-sociomantic** [comment]
As the existing Consume code is so complicated, this should be done as a complete reimplementation based on https://github.com/sociomantic-tsunami/swarm/blob/v4.x.x/src/swarm/neo/client/mixins/SuspendableRequestCore.d.

The implementation of the DHT GetAll request should provide a useful reference: [client](https://github.com/sociomantic-tsunami/dhtproto/blob/neo/src/dhtproto/client/request/internal/GetAll.d) / [node](https://github.com/sociomantic-tsunami/dhtproto/blob/neo/src/dhtproto/node/neo/request/GetAll.d).

**Comment by gavin-norman-sociomantic** [comment]
As the suspend/resume system for batch-based requests like this will be very different to the stream-based requests supported by https://github.com/sociomantic-tsunami/swarm/blob/v4.x.x/src/swarm/neo/client/mixins/SuspendableRequestCore.d, we'll need to implement a new framework on top of https://github.com/sociomantic-tsunami/swarm/blob/v4.x.x/src/swarm/neo/client/mixins/AllNodesRequestCore.d.

**Comment by gavin-norman-sociomantic** [comment]
Actually, it turns out that the request can probably be implemented in a completely synchronous way by only allowing Stop messages to be sent in-between batches.

That means there's no need to use the multi-fiber framework. https://github.com/sociomantic-tsunami/swarm/blob/v4.x.x/src/swarm/neo/client/mixins/AllNodesRequestCore.d should be useful, but that's all.

**Comment by gavin-norman-sociomantic** [comment]
Fixed by #7.
